components/GifRenderer/common/src/desktopMain/kotlin/org/jetbrains/compose/gifrenderer/DesktopGifRenderer.kt

```diff
@@ -0,0 +1,50 @@
+package org.jetbrains.compose.gifrenderer
+
+import androidx.compose.animation.core.*
+import androidx.compose.runtime.*
+import androidx.compose.ui.graphics.asImageBitmap
+import androidx.compose.ui.graphics.painter.BitmapPainter
+import androidx.compose.ui.graphics.painter.Painter
+import org.jetbrains.skija.Bitmap
+import org.jetbrains.skija.Codec
+import org.jetbrains.skija.Data
+import java.net.URL
+
+@Composable
+internal actual fun rememberGifRendererImpl(url: String): State<Painter> {
+    val codec = remember(url) {
+        println("Gif renderer for $url")
+        val bytes = URL(url).readBytes()
+        val data = Data.makeFromBytes(bytes)
+        Codec.makeFromData(data)
+    }
+
+    val images = remember(codec) {
+        (0 until codec.frameCount).map { frameIndex ->
+            val bitmap = Bitmap()
+            bitmap.allocPixels(codec.imageInfo)
+            codec.readPixels(bitmap, frameIndex)
+            BitmapPainter(bitmap.asImageBitmap())
+        }
+    }
+
+    val transition = rememberInfiniteTransition()
+    val frameIndex by transition.animateValue(
+        initialValue = 0,
+        targetValue = codec.frameCount - 1,
+        Int.VectorConverter,
+        animationSpec = infiniteRepeatable(
+            animation = keyframes {
+                durationMillis = 0
+                for ((index, frame) in codec.framesInfo.withIndex()) {
+                    index at durationMillis
+                    durationMillis += frame.duration
```

Comment:
> If the delay time of a frame is less than 20ms, mainstream browsers will set 100ms by default.
https://github.com/flutter/flutter/issues/29130#issuecomment-897732165

if use https://mixin-images.zeromesh.net/rtCKv3Mng6KY7ydK9KhKJfD7NL1QkqSsBnMd-u4sS74XId7cjLBiWPVvIrJucKOhyp0tg3YZhqTM3t5hAEoAtRU will found, only first one frame have a duration, other frame’s duration is zero.